File: TimeKeeperAndroid/app/src/main/java/be/t_ars/timekeeper/sound/WaveUtil.kt
package be.t_ars.timekeeper.sound

// by Evan X. Merz
// www.thisisnotalabel.com

// Example Wav file input and output
// this was written for educational purposes, but feel free to use it for anything you like
// as long as you credit me appropriately ("wav IO based on code by Evan Merz")

// if you catch any bugs in this, or improve upon it significantly, send me the changes
// at evan at thisisnotalabel dot com, so we can share your changes with the world

import android.content.Context
import java.io.*
import kotlin.math.floor
import kotlin.math.roundToInt
import kotlin.math.roundToLong
import kotlin.math.sin

data class WaveHeader(
    val audioFormat: Int,
    val numChannels: Int,
    val sampleRate: Long,
    val bitsPerSample: Int,
    val dataSize: Long,
)

const val SAMPLES_PER_SECOND = 44100

private const val FORMAT_PCM_8BIT = 1
private const val FORMAT_FLOAT = 3

typealias OpenFile = (String) -> InputStream

class WaveUtil(private val openFile: OpenFile) {
    constructor(context: Context) : this({ name -> context.assets.open(name) })

    /*
     WAV File Specification
     FROM http://ccrma.stanford.edu/courses/422/projects/WaveFormat/
    The canonical WAVE format starts with the RIFF header:
    0         4   ChunkID          Contains the letters "RIFF" in ASCII form
                                   (0x52494646 big-endian form).
    4         4   ChunkSize        36 + SubChunk2Size, or more precisely:
                                   4 + (8 + SubChunk1Size) + (8 + SubChunk2Size)
                                   This is the size of the rest of the chunk
                                   following this number.  This is the size of the
                                   entire file in bytes minus 8 bytes for the
                                   two fields not included in this count:
                                   ChunkID and ChunkSize.
    8         4   Format           Contains the letters "WAVE"
                                   (0x57415645 big-endian form).

    The "WAVE" format consists of two subchunks: "fmt " and "data":
    The "fmt " subchunk describes the sound data's format:
    12        4   Subchunk1ID      Contains the letters "fmt "
                                   (0x666d7420 big-endian form).
    16        4   Subchunk1Size    16 for PCM.  This is the size of the
                                   rest of the Subchunk which follows this number.
    20        2   AudioFormat      PCM = 1 (i.e. Linear quantization)
                                   Values other than 1 indicate some
                                   form of compression.
    22        2   NumChannels      Mono = 1, Stereo = 2, etc.
    24        4   SampleRate       8000, 44100, etc.
    28        4   ByteRate         == SampleRate * NumChannels * BitsPerSample/8
    32        2   BlockAlign       == NumChannels * BitsPerSample/8
                                   The number of bytes for one sample including
                                   all channels. I wonder what happens when
                                   this number isn't an integer?
    34        2   BitsPerSample    8 bits = 8, 16 bits = 16, etc.

    The "data" subchunk contains the size of the data and the actual sound:
    36        4   Subchunk2ID      Contains the letters "data"
                                   (0x64617461 big-endian form).
    40        4   Subchunk2Size    == NumSamples * NumChannels * BitsPerSample/8
                                   This is the number of bytes in the data.
                                   You can also think of this as the size
                                   of the read of the subchunk following this
                                   number.
    44        *   Data             The actual sound data.


NOTE TO READERS:

The thing that makes reading wav files tricky is that java has no unsigned types.  This means that the
binary data can't just be read and cast appropriately.  Also, we have to use larger types
than are normally necessary.

In many languages including java, an integer is represented by 4 bytes.  The issue here is
that in most languages, integers can be signed or unsigned, and in wav files the  integers
are unsigned.  So, to make sure that we can store the proper values, we have to use longs
to hold integers, and integers to hold shorts.

Then, we have to convert back when we want to save our wav data.

It's complicated, but ultimately, it just results in a few extra functions at the bottom of
this file.  Once you understand the issue, there is no reason to pay any more attention
to it.


ALSO:

This code won't read ALL wav files.  This does not use to full specification.  It just uses
a trimmed down version that most wav files adhere to.


*/
    fun generateClick(
        beepFrequency: Int,
        beepDurationMillis: Int,
        bpm: Int,
        divisionFrequency: Int,
        divisionVolume: Int,
        divisionCount: Int,
        beatCount: Int,
    ): ByteArray {
        val divisionCount = if (divisionCount in 2..7) divisionCount else 1
        val beatCount = if (beatCount in 1..7) beatCount else 1
        val samplesPerBeat = SAMPLES_PER_SECOND * 60 / bpm
        val buffer = ByteArray(beatCount * samplesPerBeat * 2) { -128 }
        (0 until beatCount).forEach { beatIndex ->
            val beatFrequency = if (beatIndex == 0) beepFrequency else divisionFrequency
            val sampleCountOffset = beatIndex * samplesPerBeat
            writeBeepInBuffer(beatFrequency, beepDurationMillis, 100, buffer, sampleCountOffset)
            (1 until divisionCount).forEach { subDivisionIndex ->
                val divisionSampleCountOffset = sampleCountOffset +
                        (samplesPerBeat.toDouble() / divisionCount.toDouble() * subDivisionIndex.toDouble()).roundToInt()
                writeBeepInBuffer(
                    divisionFrequency,
                    beepDurationMillis,
                    divisionVolume,
                    buffer,
                    divisionSampleCountOffset
                )
            }
        }
        return buffer
    }

    private fun writeBeepInBuffer(
        sinesPerSecond: Int,
        beepDurationMillis: Int,
        maxVolume: Int,
        buffer: ByteArray,
        sampleCountOffset: Int
    ) {
        val maxVolume: Double = if (maxVolume in 1..100) maxVolume.toDouble() / 100.0 else 1.0
        val desiredSampleCount =
            SAMPLES_PER_SECOND.toDouble() * beepDurationMillis.toDouble() / 1000.0
        val samplesBetweenZeroCrossings =
            SAMPLES_PER_SECOND.toDouble() / sinesPerSecond.toDouble() / 2.0
        val actualSampleCount =
            floor((desiredSampleCount / samplesBetweenZeroCrossings).roundToLong() * samplesBetweenZeroCrossings).toInt()
        val amplitudeStepPerSample =
            2.0 * Math.PI * sinesPerSecond.toDouble() / SAMPLES_PER_SECOND.toDouble()
        (0 until actualSampleCount).forEach { s ->
            val amplitude: Double = sin(s.toDouble() * amplitudeStepPerSample)
            val volumeAdjustedAmplitude: Double = amplitude * maxVolume
            val value: Double = (volumeAdjustedAmplitude + 1.0) * 255.0 / 2.0
            val byte: Byte = toByte(value.roundToInt())
            val index: Int = (sampleCountOffset + s) * 2
            if (index in buffer.indices) {
                buffer[index] = byte
            }
            if (index + 1 in buffer.indices) {
                buffer[index + 1] = byte
            }
        }
    }

    fun generateShakerLoop(bpm: Int, divisionCount: Int): ByteArray {
        val divisionCount = if (divisionCount in 1..4) divisionCount else 4
        val samplesPerBeat = SAMPLES_PER_SECOND * 60 / bpm
        val buffer = ByteArray(samplesPerBeat * 2) { -128 }
        val samples = readSamples("shakerloop", 4)
        for (i in 0 until divisionCount) {
            copyBytes(
                samples[i],
                buffer,
                (samplesPerBeat.toDouble() / divisionCount.toDouble() * i.toDouble()).roundToInt() * 2
            )
        }
        return buffer
    }

    fun generateCowbell(
        bpm: Int,
        divisionCount: Int,
        beatCount: Int,
        divisionVolume: Int
    ): ByteArray {
        val samplesPerBeat = SAMPLES_PER_SECOND * 60 / bpm
        val buffer = ByteArray(beatCount * samplesPerBeat * 2) { -128 }
        val high = readSample("high.wav")
        val low = readSample("low.wav")
        val quiet = adjustVolume(low, divisionVolume)
        copyBytes(high, buffer, 0)
        for (i in 0 until beatCount) {
            for (j in 0 until divisionCount) {
                val sample = if (j == 0)
                    if (i == 0) high else low
                else
                    quiet
                copyBytes(
                    sample,
                    buffer,
                    ((i.toDouble() + j.toDouble() / divisionCount.toDouble()) * samplesPerBeat.toDouble()).toInt() * 2
                )
            }
        }
        return buffer
    }

    private fun generateSample(bpm: Int, filename: String): ByteArray {
        val samplesPerBeat = SAMPLES_PER_SECOND * 60 / bpm
        val buffer = ByteArray(samplesPerBeat * 2) { -128 }
        val sample = readSample(filename)
        copyBytes(sample, buffer, 0)
        return buffer
    }

    fun generateCountOff(bpm: Int, beats: Int): ByteArray {
        val samplesPerBeat = SAMPLES_PER_SECOND * 60 / bpm
        val buffer = ByteArray(beats * samplesPerBeat * 2) { -128 }
        val samples = readSamples("countdown", beats)
        for (i in 0 until beats) {
            copyBytes(
                samples[i],
                buffer,
                (samplesPerBeat.toDouble() * i.toDouble()).roundToInt() * 2
            )
        }
        return buffer
    }

    private fun readSamples(prefix: String, beats: Int) =
        Array(beats) { readSample("${prefix}${it + 1}.wav") }

    private fun readSample(filename: String): ByteArray {
        openFile(filename).use { input ->
            val header = readHeader(input)
            if (header.audioFormat != FORMAT_PCM_8BIT) {
                throw IllegalArgumentException("Only 8bit PCM is supported")
            }
            if (header.bitsPerSample != 8) {
                throw IllegalArgumentException("Only 8bit PCM is supported")
            }
            if (header.sampleRate != SAMPLES_PER_SECOND.toLong()) {
                throw IllegalArgumentException("Only $SAMPLES_PER_SECOND samples per second are supported")
            }
            if (header.numChannels != 2) {
                throw IllegalArgumentException("Only 2 channels are supported")
            }
            val buffer = ByteArray(header.dataSize.toInt())
            input.read(buffer)
            return buffer
        }
    }

    companion object {
        fun readHeader(input: InputStream): WaveHeader {
            if (readWord(input) != "RIFF") {
                throw IllegalArgumentException("Expected 'RIFF'")
            }
            input.skip(4) // skip ChunkSize
            if (readWord(input) != "WAVE") {
                throw IllegalArgumentException("Expected 'WAVE'")
            }
            while (readWord(input) != "fmt ") {
                val chunkSize = read4ByteNumber(input)
                input.skip(chunkSize)
            }
            if (read4ByteNumber(input) != 16L) {
                throw IllegalArgumentException("Expected fmt size 16")
            }
            val audioFormat = read2ByteNumber(input)
            val numChannels = read2ByteNumber(input)
            val sampleRate = read4ByteNumber(input)
            input.skip(4) // skip ByteRate
            input.skip(2) // skip BlockAlign
            val bitsPerSample = read2ByteNumber(input)
            while (readWord(input) != "data") {
                val chunkSize = read4ByteNumber(input)
                input.skip(chunkSize)
            }
            val dataSize = read4ByteNumber(input)
            return WaveHeader(audioFormat, numChannels, sampleRate, bitsPerSample, dataSize)
        }

        private fun adjustVolume(input: ByteArray, volumePercentage: Int) =
            ByteArray(input.size) {
                val amplitude = input[it].toUByte().toInt() - 128
                val adjustedAmplitude =
                    (amplitude.toDouble() * volumePercentage.toDouble() / 100.0).roundToInt()
                (adjustedAmplitude + 128).toByte()
            }

        fun copyBytes(from: ByteArray, to: ByteArray, bufferOffset: Int) {
            from.indices.forEach { index ->
                val targetIndex = bufferOffset + index
                if (targetIndex in to.indices) {
                    to[targetIndex] = from[index]
                }
            }
        }

        private fun read4ByteNumber(input: InputStream): Long {
            val b1 = input.read().toLong()
            val b2 = input.read().toLong()
            val b3 = input.read().toLong()
            val b4 = input.read().toLong()
            return b1 + (b2 shl 8) + (b3 shl 16) + (b4 shl 24)
        }

        private fun read2ByteNumber(input: InputStream): Int {
            val b1 = input.read()
            val b2 = input.read()
            return b1 + (b2 shl 8)
        }

        private fun readWord(input: InputStream) =
            (0..3)
                .map {
                    input.read().toChar()
                }
                .joinToString("")

        // convert a short to a byte array
        private fun toByte(data: Int) =
            (data and 0xff).toByte()
    }
}

File: TimeKeeperIOS/TimeKeeper/WaveUtil.swift
import Foundation
import AVFoundation
import Accelerate

public struct WaveHeader {
    public let audioFormat: Int
    public let numChannels: Int
    public let sampleRate: UInt32
    public let bitsPerSample: Int
    public let dataSize: UInt32
}

public let SAMPLES_PER_SECOND: Int = 44100

private let FORMAT_PCM_8BIT: Int = 1
private let FORMAT_FLOAT: Int = 3 // retained for parity, unused

public typealias OpenFile = (_ name: String) -> InputStream?

public final class WaveUtil {
    private let openFile: OpenFile

    public init(openFile: @escaping OpenFile) {
        self.openFile = openFile
    }

    public convenience init(bundle: Bundle = .main) {
        self.init(openFile: { name in
            let url = bundle.url(
                forResource: (name as NSString).deletingPathExtension,
                withExtension: (name as NSString).pathExtension
            )
            guard let url else { return nil }
            return InputStream(url: url)
        })
    }

    // MARK: - Public API
    
    public func generateClickBuffer(
        beepFrequency: Int,
        beepDurationMillis: Int,
        bpm: UInt8,
        divisionCount: UInt8,
        beatCount: UInt8,
        mainVolume: Int,
        divisionFrequency: Int,
        divisionVolume: Int,
        stereo: Bool
    ) -> AVAudioPCMBuffer {
        let divisionCount = (2...7).contains(divisionCount) ? divisionCount : 1
        let beatCount = (1...7).contains(beatCount) ? beatCount : 1
        let samplesPerBeat = SAMPLES_PER_SECOND * 60 / max(Int(bpm), 1)
        let frameCount = UInt32(Int(beatCount) * samplesPerBeat)
        let buffer = makeBuffer(frameCount: frameCount, stereo: stereo)
        guard let channels = buffer.floatChannelData else { return buffer }

        // Silence baseline
        vDSP_vclr(channels[0], 1, vDSP_Length(frameCount))
        if (stereo) {
            vDSP_vclr(channels[1], 1, vDSP_Length(frameCount))
        }

        for beatIndex in 0..<beatCount {
            let isDownbeat = beatIndex == 0
            let freq = isDownbeat ? beepFrequency : divisionFrequency
            let sampleCountOffset = Int(beatIndex) * samplesPerBeat
            writeBeepInBuffer(
                sinesPerSecond: freq,
                beepDurationMillis: beepDurationMillis,
                volume: mainVolume,
                channels: channels,
                sampleCountOffset: sampleCountOffset,
                totalFrames: Int(frameCount),
                stereo: stereo
            )
            if divisionCount > 1 {
                for subDivisionIndex in 1..<divisionCount {
                    let divisionSampleCountOffset = Int(round((Double(samplesPerBeat) / Double(divisionCount)) * Double(subDivisionIndex))) + sampleCountOffset
                    writeBeepInBuffer(
                        sinesPerSecond: divisionFrequency,
                        beepDurationMillis: beepDurationMillis,
                        volume: mainVolume * divisionVolume / 100,
                        channels: channels,
                        sampleCountOffset: divisionSampleCountOffset,
                        totalFrames: Int(frameCount),
                        stereo: stereo
                    )
                }
            }
        }
        return buffer
    }

    public func generateShakerLoopBuffer(
        bpm: UInt8,
        divisionCount: UInt8,
        beatCount: UInt8,
        volume: Int,
        stereo: Bool
    ) -> AVAudioPCMBuffer {
        let divisionCount = (1...4).contains(divisionCount) ? divisionCount : 4
        let samplesPerBeat = SAMPLES_PER_SECOND * 60 / max(Int(bpm), 1)
        let frameCount = UInt32(Int(beatCount) * samplesPerBeat)
        let buffer = makeBuffer(frameCount: frameCount, stereo: stereo)
        guard let channels = buffer.floatChannelData else { return buffer }

        // Silence baseline
        vDSP_vclr(channels[0], 1, vDSP_Length(frameCount))
        if stereo {
            vDSP_vclr(channels[1], 1, vDSP_Length(frameCount))
        }

        let samples = readSamples(prefix: "shakerloop", beats: 4, volume: volume)
        for i in 0..<beatCount {
            for j in 0..<Int(divisionCount) {
                let offsetFrames = Int((Double(i) + Double(j) / Double(divisionCount)) * Double(samplesPerBeat))
                Self.copyBytes(from: samples[safe: j] ?? nil, channels: channels, frameOffset: offsetFrames, totalFrames: Int(frameCount), stereo: stereo)
            }
        }
        return buffer
    }

    public func generateCowbellBuffer(
        bpm: UInt8,
        divisionCount: UInt8,
        beatCount: UInt8,
        mainVolume: Int,
        divisionVolume: Int,
        stereo: Bool
    ) -> AVAudioPCMBuffer {
        let samplesPerBeat = SAMPLES_PER_SECOND * 60 / max(Int(bpm), 1)
        let frameCount = UInt32(Int(beatCount) * samplesPerBeat)
        let buffer = makeBuffer(frameCount: frameCount, stereo: stereo)
        guard let channels = buffer.floatChannelData else { return buffer }

        // Silence baseline
        vDSP_vclr(channels[0], 1, vDSP_Length(frameCount))
        if stereo {
            vDSP_vclr(channels[1], 1, vDSP_Length(frameCount))
        }

        let high = readSample(filename: "high.wav", volume: mainVolume)
        let low = readSample(filename: "low.wav", volume: mainVolume)
        var softerLow = low
        if var s = softerLow { Self.adjustVolume(&s, volumePercentage: divisionVolume); softerLow = s }

        // Downbeat
        Self.copyBytes(from: high, channels: channels, frameOffset: 0, totalFrames: Int(frameCount), stereo: stereo)
        for i in 0..<beatCount {
            for j in 0..<divisionCount {
                let sample: [UInt8]?
                if j == 0 {
                    sample = (i == 0) ? high : low
                } else {
                    sample = softerLow
                }
                let offsetFrames = Int((Double(i) + Double(j) / Double(divisionCount)) * Double(samplesPerBeat))
                Self.copyBytes(from: sample, channels: channels, frameOffset: offsetFrames, totalFrames: Int(frameCount), stereo: stereo)
            }
        }
        return buffer
    }

    public func mixCountOff(
        click: AVAudioPCMBuffer,
        bpm: UInt8,
        beats: UInt8,
        volume: Int,
        stereo: Bool
    ) -> AVAudioPCMBuffer {
        // Create a copy buffer to avoid mutating the input
        let frameCount = click.frameLength
        let out = makeBuffer(frameCount: frameCount, stereo: stereo)
        guard let inCh = click.floatChannelData, let outCh = out.floatChannelData else { return click }
        vDSP_mmov(inCh[0], outCh[0], vDSP_Length(frameCount), 1, 1, 1)
        if stereo {
            vDSP_mmov(inCh[1], outCh[1], vDSP_Length(frameCount), 1, 1, 1)
        }

        let samplesPerBeat = SAMPLES_PER_SECOND * 60 / max(Int(bpm), 1)
        let samples = readSamples(prefix: "countdown", beats: min(Int(beats), 8), volume: volume)
        for i in 0..<Int(beats) {
            let sample = samples[safe: i % max(samples.count, 1)] ?? nil
            let offsetFrames = Int(round(Double(samplesPerBeat * i)))
            Self.mixIn(channels: outCh, sample: sample, frameOffset: offsetFrames, totalFrames: Int(frameCount), stereo: stereo)
        }
        return out
    }

    // MARK: - Private helpers

    private func makeBuffer(frameCount: UInt32, stereo: Bool) -> AVAudioPCMBuffer {
        let format = AVAudioFormat(standardFormatWithSampleRate: Double(SAMPLES_PER_SECOND), channels: stereo ? 2 : 1)!
        let buf = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount)!
        buf.frameLength = frameCount
        return buf
    }

    private static func byteToFloat(_ b: UInt8) -> Float {
        return Float(Int(b) - 128) / 128.0
    }

    private func writeBeepInBuffer(
        sinesPerSecond: Int,
        beepDurationMillis: Int,
        volume: Int,
        channels: UnsafePointer<UnsafeMutablePointer<Float>>,
        sampleCountOffset: Int,
        totalFrames: Int,
        stereo: Bool
    ) {
        let maxVolume = (1...100).contains(volume) ? Float(volume) / 100.0 : 1.0
        let desiredSampleCount = Float(SAMPLES_PER_SECOND) * Float(beepDurationMillis) / 1000.0
        let samplesBetweenZeroCrossings = Float(SAMPLES_PER_SECOND) / Float(max(sinesPerSecond, 1)) / 2.0
        let actualSampleCount = Int(floor((desiredSampleCount / samplesBetweenZeroCrossings).rounded() * samplesBetweenZeroCrossings))
        if actualSampleCount <= 0 { return }
        let amplitudeStepPerSample = 2.0 * Float.pi * Float(max(sinesPerSecond, 1)) / Float(SAMPLES_PER_SECOND)
        var s = 0
        while s < actualSampleCount {
            let amp = sin(Float(s) * amplitudeStepPerSample)
            let value = amp * maxVolume
            let idx = sampleCountOffset + s
            if idx >= 0 && idx < totalFrames {
                channels[0][idx] = value
                if stereo { channels[1][idx] = value }
            }
            s += 1
        }
    }

    private static func copyBytes(
        from: [UInt8]?,
        channels: UnsafePointer<UnsafeMutablePointer<Float>>,
        frameOffset: Int,
        totalFrames: Int,
        stereo: Bool
    ) {
        guard let from = from else { return }
        let sampleFrames = from.count / 2
        var f = 0
        if stereo {
            while f < sampleFrames {
                let l8 = byteToFloat(from[2 * f])
                let r8 = byteToFloat(from[2 * f + 1])
                let idx = frameOffset + f
                if idx >= 0 && idx < totalFrames {
                    channels[0][idx] = l8
                    channels[1][idx] = r8
                }
                f += 1
            }
        } else {
            while f < sampleFrames {
                let l8 = byteToFloat(from[2 * f])
                let idx = frameOffset + f
                if idx >= 0 && idx < totalFrames {
                    channels[0][idx] = l8
                }
                f += 1
            }
        }
    }

    private static func mixIn(
        channels: UnsafePointer<UnsafeMutablePointer<Float>>,
        sample: [UInt8]?,
        frameOffset: Int,
        totalFrames: Int,
        stereo: Bool
    ) {
        guard let sample = sample else { return }
        let sampleFrames = sample.count / 2
        var f = 0
        if stereo {
            while f < sampleFrames {
                let l8 = byteToFloat(sample[2 * f])
                let r8 = byteToFloat(sample[2 * f + 1])
                let idx = frameOffset + f
                if idx >= 0 && idx < totalFrames {
                    channels[0][idx] = max(min(channels[0][idx] + l8, 1.0), -1.0)
                    channels[1][idx] = max(min(channels[1][idx] + r8, 1.0), -1.0)
                }
                f += 1
            }
        } else {
            while f < sampleFrames {
                let l8 = byteToFloat(sample[2 * f])
                let idx = frameOffset + f
                if idx >= 0 && idx < totalFrames {
                    channels[0][idx] = max(min(channels[0][idx] + l8, 1.0), -1.0)
                }
                f += 1
            }
        }
    }

    private func writeBeepInBuffer(
        sinesPerSecond: Int,
        beepDurationMillis: Int,
        volume: Int,
        buffer: inout [UInt8],
        sampleCountOffset: Int,
        stereo: Bool
    ) {
        let maxVolume = (1...100).contains(volume) ? Double(volume) / 100.0 : 1.0
        let desiredSampleCount = Double(SAMPLES_PER_SECOND) * Double(beepDurationMillis) / 1000.0
        let samplesBetweenZeroCrossings = Double(SAMPLES_PER_SECOND) / Double(max(sinesPerSecond, 1)) / 2.0
        let actualSampleCount = Int(floor((desiredSampleCount / samplesBetweenZeroCrossings).rounded() * samplesBetweenZeroCrossings))
        let amplitudeStepPerSample = 2.0 * Double.pi * Double(max(sinesPerSecond, 1)) / Double(SAMPLES_PER_SECOND)
        if actualSampleCount <= 0 { return }
        for s in 0..<actualSampleCount {
            let amplitude = sin(Double(s) * amplitudeStepPerSample)
            let volumeAdjustedAmplitude = amplitude * maxVolume
            let value = (volumeAdjustedAmplitude + 1.0) * 255.0 / 2.0
            let byteVal = UInt8(clamping: Int(value.rounded()))
            let index = (sampleCountOffset + s) * 2
            if buffer.indices.contains(index) { buffer[index] = byteVal }
            if stereo {
                let idxR = index + 1
                if buffer.indices.contains(idxR) { buffer[idxR] = byteVal }
            }
        }
    }

    private func readSamples(prefix: String, beats: Int, volume: Int) -> [[UInt8]?] {
        return (0..<beats).map { i in
            return readSample(filename: "\(prefix)\(i+1).wav", volume: volume)
        }
    }

    private func readSample(filename: String, volume: Int) -> [UInt8]? {
        guard let input = openFile(filename) else { return nil }
        input.open()
        defer { input.close() }
        do {
            let header = try Self.readHeader(input: input)
            guard header.audioFormat == FORMAT_PCM_8BIT else { throw WaveError.unsupported("Only 8bit PCM is supported") }
            guard header.bitsPerSample == 8 else { throw WaveError.unsupported("Only 8bit PCM is supported") }
            guard header.sampleRate == UInt32(SAMPLES_PER_SECOND) else { throw WaveError.unsupported("Only \(SAMPLES_PER_SECOND) samples per second are supported") }
            guard header.numChannels == 2 else { throw WaveError.unsupported("Only 2 channels are supported") }
            var buffer = [UInt8](repeating: 0, count: Int(header.dataSize))
            let readCount = input.read(&buffer, maxLength: buffer.count)
            if readCount > 0 && readCount < buffer.count {
                buffer.removeSubrange(readCount..<buffer.count)
            }
            Self.adjustVolume(&buffer, volumePercentage: volume)
            return buffer
        } catch {
            return nil
        }
    }

    // MARK: - Static utilities (parity)

    public enum WaveError: Error { case malformed(String); case unsupported(String) }

    public static func readHeader(input: InputStream) throws -> WaveHeader {
        func readWord(_ input: InputStream) throws -> String {
            var bytes = [UInt8](repeating: 0, count: 4)
            let n = input.read(&bytes, maxLength: 4)
            guard n == 4 else { throw WaveError.malformed("Unexpected EOF while reading word") }
            return String(bytes: bytes, encoding: .ascii) ?? "\u{0}\u{0}\u{0}\u{0}"
        }
        func read4(_ input: InputStream) throws -> UInt32 {
            var b = [UInt8](repeating: 0, count: 4)
            let n = input.read(&b, maxLength: 4)
            guard n == 4 else { throw WaveError.malformed("Unexpected EOF while reading 4 bytes") }
            return UInt32(b[0]) | (UInt32(b[1]) << 8) | (UInt32(b[2]) << 16) | (UInt32(b[3]) << 24)
        }
        func read2(_ input: InputStream) throws -> Int {
            var b = [UInt8](repeating: 0, count: 2)
            let n = input.read(&b, maxLength: 2)
            guard n == 2 else { throw WaveError.malformed("Unexpected EOF while reading 2 bytes") }
            return Int(UInt32(b[0]) | (UInt32(b[1]) << 8))
        }
        func skip(_ input: InputStream, count: Int) throws {
            var remaining = count
            var buf = [UInt8](repeating: 0, count: min(4096, remaining))
            while remaining > 0 {
                let n = input.read(&buf, maxLength: min(buf.count, remaining))
                if n <= 0 { throw WaveError.malformed("Unexpected EOF while skipping bytes") }
                remaining -= n
            }
        }

        guard try readWord(input) == "RIFF" else { throw WaveError.malformed("Expected 'RIFF'") }
        _ = try read4(input) // ChunkSize
        guard try readWord(input) == "WAVE" else { throw WaveError.malformed("Expected 'WAVE'") }
        var word = try readWord(input)
        while word != "fmt " {
            let chunkSize = try read4(input)
            try skip(input, count: Int(chunkSize))
            word = try readWord(input)
        }
        let fmtSize = try read4(input)
        guard fmtSize == 16 else { throw WaveError.malformed("Expected fmt size 16") }
        let audioFormat = try read2(input)
        let numChannels = try read2(input)
        let sampleRate = try read4(input)
        _ = try read4(input) // ByteRate
        _ = try read2(input) // BlockAlign
        let bitsPerSample = try read2(input)
        var word2 = try readWord(input)
        while word2 != "data" {
            let chunkSize = try read4(input)
            try skip(input, count: Int(chunkSize))
            word2 = try readWord(input)
        }
        let dataSize = try read4(input)
        return WaveHeader(audioFormat: audioFormat, numChannels: numChannels, sampleRate: sampleRate, bitsPerSample: bitsPerSample, dataSize: dataSize)
    }

    private static func adjustVolume(_ input: inout [UInt8], volumePercentage: Int) {
        guard volumePercentage < 100 else { return }
        for i in 0..<input.count {
            let amplitude = Int(input[i]) - 128
            let adjusted = Int((Double(amplitude) * Double(volumePercentage) / 100.0).rounded())
            let final = UInt8(clamping: adjusted + 128)
            input[i] = final
        }
    }

    private static func copyBytes(from: [UInt8]?, to: inout [UInt8], bufferOffset: Int, stereo: Bool) {
        guard let from = from else { return }
        let indices: [Int]
        if stereo {
            indices = Array(from.indices)
        } else {
            indices = from.indices.filter { $0 % 2 == 0 }
        }
        for index in indices {
            let targetIndex = bufferOffset + index
            if to.indices.contains(targetIndex) {
                to[targetIndex] = from[index]
            }
        }
    }

    private static func mixIn(click: inout [UInt8], sample: [UInt8]?, offset: Int, stereo: Bool) {
        guard let sample = sample, offset >= 0, offset < click.count else { return }
        let step = stereo ? 1 : 2
        let upperExclusive = min(click.count, offset + sample.count)
        var i = offset
        while i < upperExclusive {
            let a = Int(click[i]) - 128
            let b = Int(sample[i - offset]) - 128
            let c = a + b
            let clipped = max(min(c, 127), -128)
            click[i] = UInt8(clamping: clipped + 128)
            i += step
        }
    }
}

// MARK: - Safe index helper
private extension Array {
    subscript(safe index: Int) -> Element? {
        guard indices.contains(index) else { return nil }
        return self[index]
    }
}
